src/main/java/org/folio/rest/impl/ItemStorageDereferencedAPI.java

```diff
@@ -0,0 +1,198 @@
+    fields.put("permloc", "permanentLocation");
+    fields.put("temploc", "temporaryLocation");
+    fields.put("effloc", "effectiveLocation");
+    idFields.add("holdingsRecordId");
+    idFields.add("instanceId");
+    idFields.add("materialTypeId");
+    idFields.add("permanentLocationId");
+    idFields.add("temporaryLocationId");
+    idFields.add("effectiveLocationId");
+    idFields.add("permanentLoanTypeId");
+    idFields.add("temporaryLoanTypeId");
+  }
+  private String sqlQuery= "SELECT item.jsonb as item, holdingstable.jsonb as " + fields.get("holdings") + ",\n" + 
+  "instancetable.jsonb as " + fields.get("instance") + ", materialtypetable.jsonb as " + fields.get("materialtype") + ",\n" + 
+  "locationtable.jsonb as " + fields.get("permloc") + ", locationtabletemp.jsonb as " + fields.get("temploc") + ",\n" + 
+  "locationtableeffective.jsonb as " + fields.get("effloc") + ",\n"+
+  "loantable.jsonb as " + fields.get("permloan") + ", loantabletemp.jsonb as " + fields.get("temploc") + "\n"+
+  "FROM " + ITEM_TABLE + "\n"+
+  "INNER JOIN holdings_record as holdingstable on item.holdingsrecordid=holdingstable.id\n"+
+  "INNER JOIN instance as instancetable on holdingstable.instanceid=instancetable.id\n"+
+  "INNER JOIN material_type as materialtypetable on item.materialtypeid=materialtypetable.id\n"+
+  "LEFT JOIN location as locationtable on item.permanentlocationid = locationtable.id\n"+
+  "LEFT JOIN location as locationtabletemp on item.temporarylocationid = locationtabletemp.id\n"+
+  "LEFT JOIN location as locationtableeffective on item.effectivelocationid = locationtableeffective.id\n"+
+  "INNER JOIN loan_type as loantable on item.permanentloantypeid=loantable.id\n"+
+  "LEFT JOIN loan_type as loantabletemp on item.temporaryloantypeid=loantabletemp.id\n";
+
+  @Validate
+  @Override
+  public void getItemStorageDereferencedItems(
+    int offset, int limit, String query, String lang,
+    Map<String, String> okapiHeaders,
+    Handler<AsyncResult<Response>> asyncResultHandler,
+    Context vertxContext) {
+
+    List<DereferencedItem> mappedResults = new ArrayList<>();
+    String whereClause = "";
+
+    if (query != null) {
+      try {
+        CQLWrapper wrapper = new CQLWrapper(
+          new CQL2PgJSON(ITEM_TABLE + "." + JSON_COLUMN), query, limit, offset);
+        logger.info("translated SQL query: " + wrapper.toString());
+        whereClause = wrapper.toString();
+      } catch(Exception e) {
+        respondWith400Error("Invalid CQL query: " + e.getMessage(), asyncResultHandler);
+      }
+    } else {
+      whereClause = "LIMIT " + limit + " OFFSET " + offset;
+    }
+
+    PostgresClient postgresClient = PgUtil.postgresClient(vertxContext, okapiHeaders);
+    postgresClient.select(sqlQuery + whereClause, asyncResult -> {
+
+      handleSelectFailure(asyncResult, asyncResultHandler);
+
+      asyncResult.result().forEach(row -> {mappedResults.add(mapToDereferencedItem(row));});
```

Comment:
If the SQL query fails, won't this line get executed anyway (as there is no guard provided by `handleSelectFailure`) and cause a null pointer exception?

Reply:
I had assumed calling a response also terminated execution?  I haven't seen an example of an explicit break statement after those are called in the existing code-but they also tend to be used in a syntactically different way than I'm calling them here.  If not, then I'll add some kind of break statement to the error-handling function.

Reply:
> I had assumed calling a response also terminated execution?

It does not. Vert.x sending the response to the client has no effect on what further code the server executes.

This happens frequently enough that Vert.x has an error message specifically for circumstances where code attempts to write a second response after the first.

bq. I haven't seen an example of an explicit break statement after those are called in the existing code-but they also tend to be used in a syntactically different way than I'm calling them here. If not, then I'll add some kind of break statement to the error-handling function.

If the sending of the failure response is intended to be the final action, there needs to be some form of return / break to stop further execution.